plugins/webp-uploads/tests/test-picture-element.php

```diff
@@ -37,49 +38,90 @@ public function test_maybe_wrap_images_in_picture_element( bool $jpeg_and_webp,
 
 		// Create some content with the image.
-		$the_image = wp_get_attachment_image( $attachment_id, 'medium', false, array( 'class' => "wp-image-{$attachment_id}" ) );
+		$the_image = wp_get_attachment_image(
+			$attachment_id,
+			'medium',
+			false,
+			array(
+				'class' => "wp-image-{$attachment_id}",
+				'alt'   => 'Green Leaves',
+			)
+		);
 
-		// Apply the wp_content_img_tag filter.
-		$the_image = apply_filters( 'wp_content_img_tag', $the_image, 'the_content', $attachment_id );
+		$processor = new WP_HTML_Tag_Processor( $the_image );
+		$width     = 0;
+		$height    = 0;
+		$alt       = '';
+		if ( $processor->next_tag( array( 'tag_name' => 'IMG' ) ) ) {
+			$width  = (int) $processor->get_attribute( 'width' );
+			$height = (int) $processor->get_attribute( 'height' );
+			$alt    = (string) $processor->get_attribute( 'alt' );
+		}
+
+		$size_to_use                  = ( $width > 0 && $height > 0 ) ? array( $width, $height ) : 'full';
+		$image_src                    = wp_get_attachment_image_src( $attachment_id, $size_to_use );
+		list( $src, $width, $height ) = $image_src;
+		$size_array                   = array( absint( $width ), absint( $height ) );
+		$image_meta                   = wp_get_attachment_metadata( $attachment_id );
+		$sizes                        = wp_calculate_image_sizes( $size_array, $src, $image_meta, $attachment_id );
+		$image_srcset                 = wp_get_attachment_image_srcset( $attachment_id, $size_to_use );
 
-		// Check that the image is wrapped in a picture element with the correct class.
-		$picture_element = sprintf( '<picture class="wp-picture-%s" style="display: contents;">', $attachment_id );
-		if ( $expect_picture_element ) {
-			$this->assertStringContainsString( $picture_element, $the_image );
-		} else {
-			$this->assertStringNotContainsString( $picture_element, $the_image );
+		$img_src = '';
+		if ( $image_src ) {
+			$img_src = $image_src[0];
 		}
+		// Remove the last size in the srcset, as it is not needed.
+		$jpeg_srcset = substr( $image_srcset, 0, strrpos( $image_srcset, ',' ) );
+		$webp_srcset = str_replace( '.jpg', '-jpg.webp', $jpeg_srcset );
+
+		// Prepare the expected HTML by replacing placeholders with expected values.
+		$expected_html = str_replace( '{{img-width}}', $width, $expected_html );
+		$expected_html = str_replace( '{{img-height}}', $height, $expected_html );
+		$expected_html = str_replace( '{{img-src}}', $img_src, $expected_html );
+		$expected_html = str_replace( '{{img-attachment-id}}', $attachment_id, $expected_html );
+		$expected_html = str_replace( '{{img-alt}}', $alt, $expected_html );
+		$expected_html = str_replace( '{{img-srcset}}', $image_srcset, $expected_html );
+		$expected_html = str_replace( '{{img-sizes}}', $sizes, $expected_html );
+		$expected_html = str_replace( '{{jpeg-srcset}}', $jpeg_srcset, $expected_html );
+		$expected_html = str_replace( '{{webp-srcset}}', $webp_srcset, $expected_html );
```

Comment:
This can be made less verbose like as follows:

```php
$replacements = array(
  '{{img-width}}' => $width,
  '{{img-height}}' => $height,
  // ...
)
$expected_html = str_replace( array_keys( $replacements ), array_values( $replacements ), $expected_html );
```

Reply:
Yes, this seems better.